server/grpc_ac.go

```diff
@@ -0,0 +1,127 @@
+package server
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+
+	pb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/buchgr/bazel-remote/cache"
+)
+
+// ActionCache interface:
+
+func (s *grpcServer) GetActionResult(ctx context.Context,
+	req *pb.GetActionResultRequest) (*pb.ActionResult, error) {
+
+	errorPrefix := "GRPC AC GET"
+	err := s.validateHash(req.ActionDigest.Hash, errorPrefix)
+	if err != nil {
+		return nil, err
+	}
+
+	dr, _, err := s.cache.Get(cache.AC, req.ActionDigest.Hash)
+	if err != nil {
+		s.accessLogger.Printf("%s %s %s", errorPrefix, req.ActionDigest.Hash, err)
+		return nil, status.Error(codes.Unknown, err.Error())
+	}
+	if dr == nil {
+		s.accessLogger.Printf("%s %s NOT FOUND", errorPrefix, req.ActionDigest.Hash)
+		return nil, status.Error(codes.NotFound,
+			fmt.Sprintf("%s not found in AC", req.ActionDigest.Hash))
+	}
+
+	// Note: req.ActionDigest refers to the Action, not the ActionResult.
+	// There's nothing useful we can do with req.ActionDigest.SizeBytes.
+
+	data, err := ioutil.ReadAll(dr)
+	if err != nil {
+		s.accessLogger.Printf("%s %s %s", errorPrefix, req.ActionDigest.Hash, err)
+		return nil, status.Error(codes.Unknown, err.Error())
+	}
+
+	result := &pb.ActionResult{}
```

Comment:
Adding a check of this type on the client side fixed the most interesting bug I encountered when implementing our goma backend...

Unfortunately, implementing this check for the HTTP interface would break my usage of bazel-remote with our custom goma backend- we store arbitrary blobs using /ac/<key>. I could add such a check if we added a new REST endpoint which skipped these checks, eg /raw/<key> - what do you think? Then I could just rename the existing "ac" directory to "raw" in our cache and not have to start over with an empty cache.

Reply:
Sure. I think adding the check to HTTP should be done in a separate PR. We can discuss then.